pageserver/src/walreceiver.rs

```diff
@@ -182,9 +192,19 @@ fn walreceiver_main(
                 waldecoder.feed_bytes(data);
 
                 while let Some((lsn, recdata)) = waldecoder.poll_decode()? {
-                    let decoded = decode_wal_record(recdata.clone());
+                    let old_checkpoint_bytes = encode_checkpoint(checkpoint);
```

Comment:
Yes, it mean extra write in our storage per each transactions.
I agree with you that it may be too expensive.
As alternative we can use some "safe xid window" and update checkpoint only if difference with storage nextXid exceed this value. And on node restart we should always advance XID on this windows size. 
But I treat it as optimization which  can be implemented later...

Reply:
I have implemented  grouping nextXid updates. I hope that to reduce number of such updates it is enough 
to change CheckPoin.update_next_xid method in this way:
```
@@ -114,6 +115,7 @@ impl CheckPoint {
     // Next XID should be greater than new_xid.
     // Also take in account 32-bit wrap-around.
     pub fn update_next_xid(&mut self, xid: u32) {
+       let xid = xid.wrapping_add(XID_CHECKPOINT_INTERVAL - 1) & !(XID_CHECKPOINT_INTERVAL - 1);
         let full_xid = self.nextXid.value;
```
So we just store "forward aligned" xid value which reduce number of stored checkpoints but still guarantee thatXid will not be reused at next restart.  I tried XID_CHECKPOINT_INTERVAL=1024 and run 10000 pgbench transactions.
This optimization has almost no effect:

|XID_CHECKPOINT_INTERVAL|buffers=1Mb   |bufers=128Mb|wel_meta_cache|
|1024                   |2m40,696s     |2m10,745s   |1m15,832s     |
|1                      |2m43,185s     |2m4,915s    |1m15,662s     |